staging/src/k8s.io/apiserver/pkg/endpoints/filters/tenantinfo.go

```diff
@@ -28,44 +28,48 @@ import (
 // WithTenantInfo set the tenant in the requestInfo for short-path requests based on the user info from the authentication result.
 func WithTenantInfo(handler http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		ctx := req.Context()
 
-		requestor, exists := request.UserFrom(ctx)
-		if !exists {
-			responsewriters.InternalError(w, req, errors.New("The user info is missing."))
+		newReq, err := NormalizeTenant(req)
+		if err != nil {
+			responsewriters.InternalError(w, req, err)
 			return
 		}
 
-		tenantInRequestor := requestor.GetTenant()
-		if tenantInRequestor == metav1.TenantNone {
-			// temporary workaround
-			// tracking issue: https://github.com/futurewei-cloud/arktos/issues/102
-			tenantInRequestor = metav1.TenantSystem
-			//responsewriters.InternalError(w, req, errors.New(fmt.Sprintf("The tenant in the user info of %s is empty. ", requestor.GetName())))
-			//return
-		}
+		handler.ServeHTTP(w, newReq)
+	})
+}
 
-		requestInfo, exists := request.RequestInfoFrom(ctx)
-		if !exists {
-			responsewriters.InternalError(w, req, errors.New("The request info is missing."))
-			return
-		}
+// NormalizeObjectTenant sets the tenant in request info based on the user tenant.
+func NormalizeTenant(req *http.Request) (*http.Request, error) {
 
-		requestInfo.Tenant = normalizeTenant(tenantInRequestor, requestInfo.Tenant)
-		req = req.WithContext(request.WithRequestInfo(ctx, requestInfo))
+	ctx := req.Context()
 
-		handler.ServeHTTP(w, req)
-	})
-}
+	requestor, exists := request.UserFrom(ctx)
+	if !exists {
+		return nil, errors.New("The user info is missing.")
+	}
+
+	tenantInRequestor := requestor.GetTenant()
+	if tenantInRequestor == metav1.TenantNone {
+		// temporary workaround
+		// tracking issue: https://github.com/futurewei-cloud/arktos/issues/102
+		tenantInRequestor = metav1.TenantSystem
+		//responsewriters.InternalError(w, req, errors.New(fmt.Sprintf("The tenant in the user info of %s is empty. ", requestor.GetName())))
+		//return
+	}
+
+	requestInfo, exists := request.RequestInfoFrom(ctx)
+	if !exists {
+		return nil, errors.New("The request info is missing.")
+	}
 
-// normalizeObjectTenant decides what the object tenant should be based on what the user tenant is.
-func normalizeTenant(userTenant, objectTenant string) string {
 	// for a reqeust from a regular user, if the tenant in the object is empty, use the tenant from user info
 	// this is what we call "shor-path", which allows users to use traditional Kubernets API in the multi-tenancy Arktos
-	if objectTenant == metav1.TenantNone && userTenant != metav1.TenantSystem {
-		return userTenant
+	if requestInfo.Tenant == metav1.TenantNone && tenantInRequestor != metav1.TenantSystem {
```

Comment:
Suggest renaming these variables to clearly differentiate the two tenants:  resource.tenant and user.tenant.  Otherwise the logics here are hard to understand for maintainers.

Reply:
Updated.